ADDING_A_PLUGIN.md

```diff
@@ -0,0 +1,347 @@
+
+You'll probably track multiple parameters in your model: say, the accuracy, precision, and cross-entropy. Each of these will have its own `tag`. Each `tag` tracks how a particular metric or measurement evolves over time. Each run may have many `tag`s.
+
+Finally, the data for each run is stored in event files within a directory corresponding to that run. The directory name becomes the name of the run. When you launch TensorBoard, you specify a log directory (`logdir`); any runs under this directory will be picked up and loaded into TensorBoard.
+
+[`add_summary`]: https://www.tensorflow.org/api_docs/python/tf/summary/FileWriter#add_summary
+
+
+## Plugins: Three Easy Pieces
+
+Now that we have the background out of the way, we can talk about plugins! Each plugin consists of three basic parts:
+
+  - The **API layer** is how users of your plugins will write data that your plugin can read.
+  - The **Backend layer** is where you write Python code that does post-processing of your data, and serves the data to your plugin frontend in the browser.
+  - The **Frontend layer** is where your custom visualization lives.
+
+### API layer: How the plugin gets data
+
+Users of your plugin will use your API to emit protocol bufers containing the data that your plugin will visualize. In general, you should expose two ways of doing this: via a TensorFlow summary op, and by directly creating a protocol buffer.
+
+For example, let's create a `Greeter` plugin. It provides a greeter `op` method, which creates a TensorFlow op that emits a string-tensor summary. It also provides a `pb` function that creates this directly, outside of a TensorFlow context. Standard TensorBoard plugins should follow this API.
+
+Here's the code; we'll explain it below:
+
+```python
+"""This module provides summaries for the Greeter plugin."""
+
+import tensorflow as tf
+
+
+PLUGIN_NAME = 'greeter'
+
+
+def op(name,
+       guest,
+       display_name=None,
+       description=None,
+       collections=None):
+  """Create a TensorFlow summary op to greet the given guest.
+
+  Arguments:
+    name: A name for this summary operation.
+    guest: A rank-0 string `Tensor`.
+    display_name: If set, will be used as the display name
+      in TensorBoard. Defaults to `name`.
+    description: A longform readable description of the summary data.
+      Markdown is supported.
+    collections: Which TensorFlow graph collections to add the summary
+      op to. Defaults to `['summaries']`. Can usually be ignored.
+  """
+
+  # The `name` argument is used to generate the summary op node name.
+  # That node name will also involve the TensorFlow name scope.
+  # By having the display_name default to the name argument, we make
+  # the TensorBoard display clearer.
+  if display_name is None:
+    display_name = name
+  with tf.name_scope(name), \
```

Comment:
Good catch. This is my mistake. (Originally we had a `tf.control_dependencies` here but removed it for simplicity.) This kind of thing is exactly why I'd like the code to be checked in. :-)